storage/src/index/storage.rs

```diff
@@ -142,201 +155,177 @@ impl<'a, T: Translator, V> Cursor<'a, T, V> {
                 });
 
-                // Set current next to be the new record.
-                self.next = Some(new);
-
-                // Now that the current item is in next, we need to update
-                // the phase.
-                self.phase = Phase::Next;
+                // Set the phase to the new record.
+                self.phase = Phase::PostInsert(new);
                 self.collisions.inc();
             }
-            Phase::Next => {
-                // Take next
-                let mut next = self.next.take().unwrap();
-                let next_next = next.next.take();
+            Phase::Next(mut current) => {
+                // Take next.
+                let next = current.next.take();
 
-                // Add next to the past list.
-                self.past_push(next);
+                // Add current to the past list.
+                self.past_push(current);
 
-                // Create a new record that points to next's next.
-                let new = Box::new(Record {
-                    value: v,
-                    next: next_next,
-                });
-                self.next = Some(new);
+                // Create a new record that points to the next's next.
+                let new = Box::new(Record { value: v, next });
+                self.phase = Phase::PostInsert(new);
                 self.collisions.inc();
             }
             Phase::Done => {
-                // If entry is deleted, we need to update it.
-                if self.entry_deleted {
-                    self.entry_deleted = false;
-                    self.entry.as_mut().unwrap().get_mut().value = v;
-                    return;
-                }
-
-                // If not, we should add to past.
+                // If we are done, we need to create a new record and
+                // immediately push it to the past list.
                 let new = Box::new(Record {
                     value: v,
                     next: None,
                 });
                 self.past_push(new);
                 self.collisions.inc();
             }
+            Phase::EntryDeleted => {
+                // If entry is deleted, we need to update it.
+                self.entry.as_mut().unwrap().get_mut().value = v;
+            }
+            Phase::PostDeleteEntry | Phase::PostDeleteNext(_) | Phase::PostInsert(_) => {
+                unreachable!("{MUST_CALL_NEXT}")
+            }
         }
     }
 
     /// Deletes the current value, adjusting the list structure.
     ///
     /// Increments the `pruned` counter to track removals.
     pub fn delete(&mut self) {
-        assert!(!self.next_required, "{MUST_CALL_NEXT}");
-        self.next_required = true;
-        self.last_deleted = true;
         self.pruned.inc();
-        match self.phase {
-            Phase::Initial => {
-                unreachable!("{MUST_CALL_NEXT}")
-            }
+        match std::mem::replace(&mut self.phase, Phase::Done) {
```

Comment:
I know we weren't doing this before, but should we be decrementing the collisions counter whenever we delete a value and there is still some other value remaining?